File: solutions/python/trees/binary-tree-level-order-traversal-ii.py
# LeetCode 107 — Binary Tree Level Order Traversal II (Medium)
# Category: Trees · Approach: BFS
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/

from collections import deque

def level_order_bottom(root) -> list[list[int]]:
    if not root:
        return []
    res, q = [], deque([root])
    while q:
        level = []
        for _ in range(len(q)):
            node = q.popleft()
            level.append(node.val)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        res.append(level)
    return res[::-1]

File: solutions/typescript/trees/binary-tree-level-order-traversal-ii.ts
// LeetCode 107 — Binary Tree Level Order Traversal II (Medium)
// Category: Trees · Approach: BFS
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/

function levelOrderBottom(root: TreeNode | null): number[][] {
  if (!root) return [];
  const res: number[][] = [];
  let q: TreeNode[] = [root];
  while (q.length) {
    const vals: number[] = [];
    const next: TreeNode[] = [];
    for (const n of q) {
      vals.push(n.val);
      if (n.left) next.push(n.left);
      if (n.right) next.push(n.right);
    }
    res.push(vals);
    q = next;
  }
  return res.reverse();
}
